### PR TITLE
pull docker images which are not already present

### PR DESCRIPTION
**Pull request by steiler**
Docker images are not pulled by default. If e.g. the henderiw/client-alpine:1.0.0 image is not present the deploy task will fail.
This branch will try downloading the container images which are missing on the system.
This would later on also help when SRL is available via a registry. But if SRL is now already loaded everything is fine.

### CODE REVIEW

File: clab/docker.go [hunk 1]
```diff
@@ -179,6 +187,45 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
...
+		} else if strings.Count(node.Image, "/") == 1 {
+			canonicalImageName = "docker.io/" + node.Image
+		}
+		log.Infof("Pulling %s Docker image", node.Image)
+		reader, err := c.DockerClient.ImagePull(ctx, canonicalImageName, types.ImagePullOptions{})
+		// must read from reader, otherwise image is not properly pulled
+		io.Copy(ioutil.Discard, reader)
+		log.Infof("Done pulling %s", canonicalImageName)
+		if err != nil {
+			panic(err)
```
Comment (hellt) [inline clab/docker.go:221]: lets return an error here instead of panicing

File: clab/docker.go [hunk 2]
```diff
@@ -179,6 +187,45 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
...
+	images, err := c.DockerClient.ImageList(ctx, ilo)
+	_ = images
+	// Image doesn't exist, need to pull it
+	if len(images) <= 0 {
+		// might need canonical name e.g.
+		//    -> alpine == docker.io/library/alpine
+		//    -> foo/bar == docker.io/foo/bar
+		//    -> docker.elastic.co/elasticsearch/elasticsearch == docker.elastic.co/elasticsearch/elasticsearch
+		canonicalImageName := node.Image
+		if strings.Count(node.Image, "/") == 0 {
```
Comment (hellt) [inline clab/docker.go:210]: lets rewrite `if/else if` to a more canonical `switch/case`?

File: clab/docker.go [hunk 3]
```diff
@@ -179,6 +187,45 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
...
+
+	ilo := types.ImageListOptions{
+		All:     false,
+		Filters: filter,
+	}
+
+	log.Debugf("Looking up %s Docker image", node.Image)
+
+	images, err := c.DockerClient.ImageList(ctx, ilo)
+	_ = images
```
Comment (hellt) [inline clab/docker.go:202]: why is this needed?

File: clab/docker.go [hunk 4]
```diff
@@ -179,6 +187,45 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
...
+	// Image doesn't exist, need to pull it
+	if len(images) <= 0 {
```
Comment (hellt) [inline clab/docker.go:204]: can length over images return a negative?

**Review by hellt — COMMENTED** [review]
thanks @steiler 
I have a few questions/suggestions

File: clab/docker.go (hunk 1 above)
Comment (steiler) [reply clab/docker.go:221]: will do

File: clab/docker.go (hunk 2 above)
Comment (steiler) [reply clab/docker.go:210]: will do

File: clab/docker.go (hunk 3 above)
Comment (steiler) [reply clab/docker.go:202]: L202 is not needed any more, you're right I'm now using that value.

File: clab/docker.go (hunk 4 above)
Comment (steiler) [reply clab/docker.go:204]: will adjust

File: clab/docker.go [hunk 5]
```diff
@@ -179,6 +187,52 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
...
+		//    -> docker.elastic.co/elasticsearch/elasticsearch == docker.elastic.co/elasticsearch/elasticsearch
+		canonicalImageName := node.Image
+		slashCount := strings.Count(node.Image, "/")
+
+		switch slashCount {
+		case 0:
+			canonicalImageName = "docker.io/library/" + node.Image
+		case 1:
+			canonicalImageName = "docker.io/" + node.Image
+		}
```
Comment (hellt) [inline clab/docker.go:220]: when I looked into SDK examples I saw that they don't use the canonical name https://docs.docker.com/engine/api/sdk/examples/#pull-an-image

can you test if that is needed? My assumption is that if you specify simply `alpine` it will still do the pull

File: clab/docker.go [hunk 6]
```diff
@@ -179,6 +187,52 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
...
+
+	log.Debugf("Looking up %s Docker image", node.Image)
+
+	images, err := c.DockerClient.ImageList(ctx, ilo)
+	if err != nil {
+		return err
+	}
+
+	// Image doesn't exist, need to pull it
+	if len(images) == 0 {
```
Comment (hellt) [inline clab/docker.go:207]: lets keep the "happy path" on the first indentation level, meaning that the structure of the code should be revered like that:

```go
if len(images) > 0:
    log.Debugf("Image %s present, skip pulling", node.Image)
    return nil

// here goes the pulling logic on the same indentation level
```

File: clab/docker.go [hunk 7]
```diff
@@ -179,6 +187,52 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
...
+
+		switch slashCount {
+		case 0:
+			canonicalImageName = "docker.io/library/" + node.Image
+		case 1:
+			canonicalImageName = "docker.io/" + node.Image
+		}
+
+		log.Infof("Pulling %s Docker image", node.Image)
+		reader, err := c.DockerClient.ImagePull(ctx, canonicalImageName, types.ImagePullOptions{})
```
Comment (hellt) [inline clab/docker.go:223]: add `defer reader.Close()` after you created a readcloser on L223

File: clab/docker.go [hunk 8]
```diff
@@ -179,6 +187,52 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
...
+		// must read from reader, otherwise image is not properly pulled
+		io.Copy(ioutil.Discard, reader)
```
Comment (hellt) [inline clab/docker.go:225]: I also think maybe you should read to `os.Stdout`? I think that way you will be able to show the progress of image pulling to a user during the lab instantiation

**Comment by hellt** [comment]
@steiler to finish this off, can you change this line https://github.com/srl-wim/container-lab/blame/master/docs/install.md#L6 to mention that only images which can't be pulled from public registries have to be pulled upfront?

File: clab/docker.go (hunk 5 above)
Comment (steiler) [reply clab/docker.go:220]: I did and I was shocked myself.

```
Start - alpine
repository name must be canonical
Done - alpine

Start - library/alpine
repository name must be canonical
Done - library/alpine

Start - docker.io/library/alpine
{"status":"Pulling from library/alpine","id":"latest"}
...
{"status":"Pull complete","progressDetail":{},"id":"df20fa9351a1"}
Done - docker.io/library/alpine
```

File: clab/docker.go (hunk 6 above)
Comment (steiler) [reply clab/docker.go:207]: I dislike this approach.
The happy path in my view is when the last statement of a function is reached without any early exiting via any return other then the last line of a function... (exceptions exist of course)
However I'll change order and have the image exists atop.

File: clab/docker.go (hunk 7 above)
Comment (steiler) [reply clab/docker.go:223]: Thanks, will do!

File: clab/docker.go (hunk 5 above)
Comment (hellt) [reply clab/docker.go:220]: Interesting, I see that we use https://github.com/srl-wim/container-lab/blob/f11586e6916cb0bf0600633073e1bb001a6f4e4b/clab/clab.go#L8 as a docker api package, although this is a fork from `moby/moby`
the official Go docs actually say to use the `github.com/docker/docker/client` package

I wonder, can you swap it and give it a try? Maybe the upstream Go SDK is a more polished version that we should rather use?

/cc @karimra

File: clab/docker.go (hunk 8 above)
Comment (steiler) [reply clab/docker.go:225]: Agree, but after having seen the output I've decided to discard.

And that's just the output for the tiny alpine image...
```
Start -  docker.io/library/alpine
{"status":"Pulling from library/alpine","id":"latest"}

{"status":"Pulling fs layer","progressDetail":{},"id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":29170,"total":2797541},"progress":"[\u003e                                                  ]  29.17kB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":425216,"total":2797541},"progress":"[=======\u003e                                           ]  425.2kB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":781568,"total":2797541},"progress":"[=============\u003e                                     ]  781.6kB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":1207552,"total":2797541},"progress":"[=====================\u003e                             ]  1.208MB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":1604061,"total":2797541},"progress":"[============================\u003e                      ]  1.604MB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":2071005,"total":2797541},"progress":"[=====================================\u003e             ]  2.071MB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":2464221,"total":2797541},"progress":"[============================================\u003e      ]  2.464MB/2.798MB","id":"df20fa9351a1"}

{"status":"Verifying Checksum","progressDetail":{},"id":"df20fa9351a1"}

{"status":"Download complete","progressDetail":{},"id":"df20fa9351a1"}

{"status":"Extracting","progressDetail":{"current":32768,"total":2797541},"progress":"[\u003e                                                  ]  32.77kB/2.798MB","id":"df20fa9351a1"}

{"status":"Extracting","progressDetail":{"current":2064384,"total":2797541},"progress":"[====================================\u003e              ]  2.064MB/2.798MB","id":"df20fa9351a1"}

{"status":"Extracting","progressDetail":{"current":2797541,"total":2797541},"progress":"[==================================================\u003e]  2.798MB/2.798MB","id":"df20fa9351a1"}

{"status":"Pull complete","progressDetail":{},"id":"df20fa9351a1"}

{"status":"Digest: sha256:185518070891758909c9f839cf4ca393ee977ac378609f700f60a771a2dfe321"}

{"status":"Status: Downloaded newer image for alpine:latest"}

Done -  docker.io/library/alpine
```

File: clab/docker.go (hunk 6 above)
Comment (hellt) [reply clab/docker.go:207]: that is what is considered an idiomatic Go (left aligned happy path with early returns), so its good to adhere to what community adopted, that will make code look consistent

File: clab/docker.go (hunk 8 above)
Comment (hellt) [reply clab/docker.go:225]: valid point. especially considering that we will pull in parallel quite soon.

File: clab/docker.go (hunk 5 above)
Comment (karimra) [reply clab/docker.go:220]: This was changed because of this issue #23

Comment (hellt) [reply clab/docker.go:220]: yes, I remember #23, it just maybe it was not the right pointer =)

File: clab/docker.go (hunk 6 above)
Comment (karimra) [reply clab/docker.go:207]: Talking about idiomatic go: https://go-proverbs.github.io/
Last one is "don't panic" includes both panic statement and log.Fatalx

Comment (steiler) [reply clab/docker.go]: > that is what is considered an idiomatic Go (left aligned happy path with early returns), so its good to adhere to what community adopted, that will make code look consistent

Well if that's the case ... who am I to complain, right?
It's changed!

File: clab/docker.go (hunk 5 above)
Comment (steiler) [reply clab/docker.go]: The official `github.com/docker/docker/client` works fine as well and doesn't break anything, but the canonical name logic is still required.

**Comment by hellt** [comment]
lgtm
